adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java

```diff
@@ -0,0 +1,141 @@
+import io.nosqlbench.adapter.jdbc.exceptions.JDBCPgVectorException;
+import io.nosqlbench.adapter.jdbc.utils.JDBCPgVector;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+
+public abstract class JDBCDMLOp extends JDBCOp {
+    private static final Logger LOGGER = LogManager.getLogger(JDBCDMLOp.class);
+    protected final boolean isReadStmt;
+    // If the passed-in statement string has placeholder '?', it is a prepared statement.
+    protected final boolean isPreparedStmt;
+    protected final String pStmtSqlStr;
+    protected final List<Object> pStmtValList;
+
+    protected static ThreadLocal<Statement> jdbcStmtTL = ThreadLocal.withInitial(() -> null);
+
+    public JDBCDMLOp(JDBCSpace jdbcSpace,
+                     boolean isReadStmt,
+                     String pStmtSqlStr,
+                     List<Object> pStmtValList) {
+        super(jdbcSpace);
+
+        this.isReadStmt = isReadStmt;
+        this.pStmtSqlStr = pStmtSqlStr;
+        this.pStmtValList = pStmtValList;
+        this.isPreparedStmt = StringUtils.contains(pStmtSqlStr, "?");
+
+        // NOTE:
+        // - The write DML statement MUST use prepared statement!
+        // - The read DML statement can use either a prepared or a regular statement.
+        if (!isReadStmt && !isPreparedStmt) {
+            throw new JDBCAdapterInvalidParamException(
+                "Write DML statement must use prepared statement format!");
+        }
+
+        if (isPreparedStmt) {
+            int expectedFiledCnt = StringUtils.countMatches(pStmtSqlStr, "?");
+            int actualFieldCnt = pStmtValList.size();
+            if (expectedFiledCnt != actualFieldCnt) {
+                throw new JDBCAdapterUnexpectedException(
+                    "Provided value count (" + actualFieldCnt
+                        + ") doesn't match the expected field count (" + expectedFiledCnt
+                        + ") for the prepared statement: \"" + pStmtSqlStr + "\""
+                );
+            }
+        }
+    }
+
+    // Only applicable to a prepared statement
+    protected PreparedStatement setPrepStmtValues(PreparedStatement stmt, List<Object> valList) {
+        assert (stmt != null);
+
+        for (int i=0; i<valList.size(); i++) {
+            int fieldIdx = i + 1;
+            String inputFieldVal = ((String)valList.get(i)).trim();
+
+            try {
+                Object fieldValObj = inputFieldVal;
+
+                // If the 'fieldVal' is a string like "[<float_num_1>, <float_num_2>, ... <float_num_n>]" format,
+                // convert it to the Vector object
+                if ( inputFieldVal.startsWith("[") && inputFieldVal.endsWith("]") ) {
+                    JDBCPgVector vector = new JDBCPgVector();
```

Comment:
#nit should we throw this declaration `JDBCPgVector vector` before the _for_ loop to reduce multiple object creation?

Reply:
fixed